util/logger.go

```diff
@@ -61,18 +63,59 @@ func getLoggingLevelFromString(level string, defaultLevel logging.Level) (loggin
 
 // LogCatcher traps messsage containing logging level [LOGLEVEL] and redirect them to the logging system
 type LogCatcher struct {
-	Writer io.Writer
-	Logger *logging.Logger
+	Writer    io.Writer
+	Logger    *logging.Logger
+	remaining string
 }
 
-// This methods intercepts every message written to stderr and determines if a logging
+// Close implements io.Closer
+func (catcher *LogCatcher) Close() error {
+	if catcher.remaining != "" {
+		catcher.Write([]byte("\n"))
+	}
+	return nil
+}
+
+func (catcher *LogCatcher) write(s string) (int, error) {
+	_, err := catcher.Writer.Write([]byte(s))
+	return len(s), err
+}
+
+// This methods intercepts every message written to stream and determines if a logging
 // function should be used.
-func (catcher LogCatcher) Write(p []byte) (n int, err error) {
+func (catcher *LogCatcher) Write(p []byte) (n int, err error) {
+	var buffer string
+	if catcher.remaining != "" {
+		n -= len(catcher.remaining)
+		buffer = catcher.remaining + string(p)
+		catcher.remaining = ""
+	} else {
+		buffer = string(p)
+	}
+
+	var errArray errors.Array
 	if logMessage == nil {
 		initLogCatcher()
 	}
-	if matches := logMessage.FindSubmatch(p); matches != nil {
-		if level, err := logging.LogLevel(string(matches[1])); err == nil {
+	if lastCR := strings.LastIndex(buffer, "\n"); lastCR >= 0 {
+		catcher.remaining = buffer[lastCR+1:]
+		buffer = buffer[:lastCR+1]
+		n += len(catcher.remaining)
```

Comment:
I don't understand all the operations with the `n` variable. Shouldn't you just return the `n` returned from the underlying Writer. From here: `_, err := catcher.Writer.Write([]byte(s))` from the function `func (catcher *LogCatcher) write(s string) (int, error) {`

Reply:
Its because the Write call must return the exact length of the supplied buffer, no more, no less. It would be simpler to just return `len(p)`, but that doesn't ensure that the function is working properly and not forget some data.

We also have to deal with the fact that the supplied buffer could be cut anywhere, it's the purpose of the `catcher.remaining`. That way, we ensure that we always deal with complete line.

The previous version was catastrophic in that matter 🙂.